pallets/registrar/src/lib.rs

```diff
@@ -620,6 +586,116 @@ pub mod pallet {
+
+            // Verify manager signature
+            let signature_msg = Self::relay_signature_msg(para_id, &account, relay_storage_root);
+            if !manager_signature.verify(&*signature_msg, &relay_manager) {
+                return Err(Error::<T>::InvalidRelayManagerSignature.into());
+            }
+
+            Self::do_register(account, para_id, genesis_data)?;
+            // Insert parathread params
+            if let Some(parathread_params) = parathread_params {
+                ParathreadParams::<T>::insert(para_id, parathread_params);
+            }
+            Self::deposit_event(Event::ParaIdRegistered { para_id });
+
+            Ok(())
+        }
+
+        /// Deregister a parachain that no longer exists in the relay chain. The origin of this
+        /// extrinsic will be rewarded with the parachain deposit.
+        #[pallet::call_index(10)]
+        #[pallet::weight(T::WeightInfo::deregister_with_relay_proof_immediate(
+            T::MaxGenesisDataSize::get(),
+            T::MaxLengthParaIds::get()
+        ).max(T::WeightInfo::deregister_with_relay_proof_scheduled(
+            T::MaxGenesisDataSize::get(),
+            T::MaxLengthParaIds::get()
+        )))]
+        pub fn deregister_with_relay_proof(
+            origin: OriginFor<T>,
+            para_id: ParaId,
+            relay_proof_block_number: u32,
+            relay_storage_proof: sp_trie::StorageProof,
+        ) -> DispatchResult {
+            let account = T::RegisterWithRelayProofOrigin::ensure_origin(origin)?;
+
+            let relay_storage_root =
+                T::RelayStorageRootProvider::get_relay_storage_root(relay_proof_block_number)
+                    .ok_or(Error::<T>::RelayStorageRootNotFound)?;
+            let relay_state_proof =
+                GenericStateProof::<cumulus_primitives_core::relay_chain::Block>::new(
+                    relay_storage_root,
+                    relay_storage_proof,
+                )
+                .map_err(|_| Error::<T>::InvalidRelayStorageProof)?;
+
+            let bytes = para_id.twox_64_concat();
+            let key = [REGISTRAR_PARAS_INDEX, bytes.as_slice()].concat();
+            // TODO: we don't even need to decode the value, only check if it exists
+            // Need to add exists_storage method to dancekit
+            let relay_para_info = relay_state_proof
+                .read_optional_entry::<ParaInfo<
+                    cumulus_primitives_core::relay_chain::AccountId,
+                    cumulus_primitives_core::relay_chain::Balance,
+                >>(key.as_slice())
+                .map_err(|_| Error::<T>::InvalidRelayStorageProof)?;
+            if relay_para_info.is_some() {
+                return Err(Error::<T>::ParaStillExistsInRelay.into());
+            }
+
+            // Take the deposit immediately and give it to origin account
+            if let Some(asset_info) = RegistrarDeposit::<T>::take(para_id) {
```

Comment:
This needs to come after the deregister. Because:

- if the deregister has already been made by the manager, then it is very likely that the deregistration in the relay has not finished yet. Therefore this slashing condition will apply. And the deposit will still have not been unreserved because we do it when the deregistration completes.

- In contrast if we do it after the deregister, the deregister should error if you are trying to deregister something that is already in the deregistering process.

What do you think? If agreed, can we create a test for this?

Reply:
Nevermind, I am still with the past mentality in which extrinsics where not transactional. Forget about it :)

Reply:
The deposit needs to be taken before, because there is an edge case in deregister that allows the deposit to be removed immediately, so it will not exist if we try to take the deposit after the call to deregister. But I should add tests.